worker/tracing/tracer.go

```diff
@@ -0,0 +1,213 @@
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+		defer cancel()
+
+		if err := t.client.Stop(ctx); err != nil {
+			t.logger.Infof("failed to stop client: %v", err)
+		}
+
+		if err := t.clientProvider.Shutdown(ctx); err != nil {
+			t.logger.Infof("failed to shutdown provider: %v", err)
+		}
+	}()
+
+	<-t.tomb.Dying()
+	return tomb.ErrDying
+}
+
+// scopedContext returns a context that is in the scope of the worker lifetime.
+// It returns a cancellable context that is cancelled when the action has
+// completed.
+func (w *tracer) scopedContext(ctx context.Context) (context.Context, context.CancelFunc) {
+	ctx, cancel := context.WithCancel(ctx)
+	return w.tomb.Context(ctx), cancel
+}
+
+func newClient(ctx context.Context, namespace, endpoint string, insecureSkipVerify bool) (otlptrace.Client, *sdktrace.TracerProvider, trace.Tracer, error) {
+	options := []otlptracegrpc.Option{
+		otlptracegrpc.WithEndpoint(endpoint),
+	}
+	if insecureSkipVerify {
+		options = append(options, otlptracegrpc.WithInsecure())
+	}
+	client := otlptracegrpc.NewClient(options...)
+	exporter, err := otlptrace.New(ctx, client)
+	if err != nil {
+		return nil, nil, nil, errors.Trace(err)
+	}
+
+	tp := sdktrace.NewTracerProvider(
+		sdktrace.WithBatcher(exporter),
+		sdktrace.WithResource(newResource(namespace)),
+	)
+
+	return client, tp, tp.Tracer(namespace), nil
+}
+
+func newResource(namespace string) *resource.Resource {
+	return resource.NewWithAttributes(
+		semconv.SchemaURL,
+		semconv.ServiceName(namespace),
+		semconv.ServiceVersion(version.Current.String()),
+	)
+}
+
+type managedSpan struct {
+	span               trace.Span
+	cancel             context.CancelFunc
+	stackTracesEnabled bool
+}
+
+// AddEvent will record an event for this span. This is a manual mechanism
+// for recording an event, it is useful to log information about what
+// happened during the lifetime of a span.
+func (s *managedSpan) AddEvent(message string, attrs ...coretracing.Attribute) {
+	// According to the docs, events can only be recorded if the span
+	// is being recorded.
+	if s.span.IsRecording() {
+		return
+	}
+
+	s.span.AddEvent(message, trace.WithAttributes(attributes(attrs)...))
+}
+
+// RecordError will record err as an exception span event for this span. This
+// also sets the span status to Error if the error is not nil.
+func (s *managedSpan) RecordError(err error, attrs ...coretracing.Attribute) {
+	if err == nil {
+		return
+	}
+
+	s.span.RecordError(err, trace.WithAttributes(attributes(attrs)...))
+	s.span.SetStatus(codes.Error, err.Error())
+}
+
+func (s *managedSpan) End(attrs ...coretracing.Attribute) {
+	defer s.cancel()
+
+	s.span.SetAttributes(attributes(attrs)...)
+	s.span.End(trace.WithStackTrace(s.stackTracesEnabled))
+}
+
+func attributes(attrs []coretracing.Attribute) []attribute.KeyValue {
+	kv := make([]attribute.KeyValue, len(attrs))
+	for _, attr := range attrs {
+		kv = append(kv, attribute.String(attr.Key(), attr.Value()))
```

Comment:
Shame they are not using generics; we could too.

Reply:
Indeed